File: Java/Maths/ArmstrongNumber.java
/*
Approach:
1. Read the input number from the user.
2. Check if the number is an Armstrong number by calling the `isArmstrongNumber` function.
3. In the `isArmstrongNumber` function, count the number of digits in the given number and store it in `numberOfDigits`.
4. Calculate the Armstrong number by iterating through each digit of the number and adding the digit raised to the power of `numberOfDigits` to the `result` variable.
5. Finally, compare the calculated `result` with the original number. If they are equal, return `true`, indicating that the number is an Armstrong number.

Time Complexity: O(log10(n)). The number of digits in a number `n` can be determined in logarithmic time.

Space Complexity: O(1). Constant space is used throughout the program, irrespective of the input.

Sample Input:
Enter a number: 153

Sample Output:
153 is an Armstrong number.
*/

//program in Java to check if a number is an Armstrong number:

import java.util.Scanner;

public class ArmstrongNumber {

    public static void main(String[] args) {
        
        // Create a Scanner object to read input from the user
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter a number: ");
        
        // Read the input number
        int number = scanner.nextInt();
        scanner.close();
        
        // Check if the number is an Armstrong number
        if(isArmstrongNumber(number)){
            System.out.println(number + " is an Armstrong number.");
        } else {
            System.out.println(number + " is not an Armstrong number.");
        }
    }
    
    // Function to check if a number is an Armstrong number
    public static boolean isArmstrongNumber(int number) {
        int originalNumber, remainder, result = 0, numberOfDigits = 0;
        originalNumber = number;
        
        // Count the number of digits in the given number
        while (originalNumber != 0) {
            originalNumber /= 10;
            ++numberOfDigits;
        }
        
        originalNumber = number;
        
        // Calculate the Armstrong number
        while (originalNumber != 0) {
            remainder = originalNumber % 10;
            result += Math.pow(remainder, numberOfDigits);
            originalNumber /= 10;
        }
        
        // Check if the calculated result is equal to the original number
        if (result == number)
            return true;
        
        return false;
    }
}

File: Python/Maths/ArmstrongNumber.py
'''
Explanation of approach:
1. The program defines a function `is_armstrong_number()` to check if a number is an Armstrong number.
2. The function takes a number as input.
3. The number is converted to a string using `str()` and the length of the number is calculated using `len()` function.
4. Using a list comprehension, the sum of each digit raised to the power of the length is calculated.
5. The calculated sum is compared with the original number to determine if it is an Armstrong number.
6. The function returns `True` if the number is an Armstrong number, else it returns `False`.
7. A sample input number is given (`153` in this case) and the function is called to check if it is an Armstrong number.
8. The result is printed on the console.

Time Complexity: In this program, the time complexity is O(n), where n is the number of digits in the given number.

Space Complexity: The space complexity of this program is O(1), as the space used does not increase with the input size.

Sample Input: 153
Sample Output: 153 is an Armstrong number
'''

def is_armstrong_number(num):
    #Check if a number is an Armstrong number

    # Convert the number to string to calculate length
    num_str = str(num)

    # Calculate the length of the number
    num_len = len(num_str)

    # Calculate the sum of the digits raised to the power of the length
    armstrong_sum = sum(int(digit) ** num_len for digit in num_str)

    # Check if the sum is equal to the original number
    if armstrong_sum == num:
        return True
    else:
        return False


# Sample input
number = 153

# Check if the number is an Armstrong number
if is_armstrong_number(number):
    print(number, "is an Armstrong number")
else:
    print(number, "is not an Armstrong number")

